From d0e5729c16c85ec731c9e1fbc49d072380680bc3 Mon Sep 17 00:00:00 2001
From: Sachin Kohli <quic_sachkohl@quicinc.com>
Date: Tue, 19 Dec 2023 10:11:20 +0530
Subject: [PATCH 351/500] vxlan: Get the remote IP address of the given client.

Given a mac address of the client and the netdevice,
return the remote IP address used by the client.

Change-Id: I3b241ea0faec61205054d8a41ed45bc10c706031
Signed-off-by: Sneha Maganahalli <quic_smaganah@quicinc.com>
Signed-off-by: Vishnu Vardhan Bantanahal <quic_vishvard@quicinc.com>
---
 drivers/net/vxlan/vxlan_core.c | 25 +++++++++++++++++++++++++
 include/net/vxlan.h            |  4 ++++
 2 files changed, 29 insertions(+)

diff --git a/drivers/net/vxlan/vxlan_core.c b/drivers/net/vxlan/vxlan_core.c
index 41612a21dbb1..53186bdc5513 100644
--- a/drivers/net/vxlan/vxlan_core.c
+++ b/drivers/net/vxlan/vxlan_core.c
@@ -460,6 +460,31 @@ static struct vxlan_fdb *vxlan_find_mac(struct vxlan_dev *vxlan,
 	return f;
 }
 
+/* Get the remote IP address of the given client */
+int vxlan_find_remote_ip(struct vxlan_dev *vxlan,
+				const u8 *mac, __be32 vni, union vxlan_addr *rip)
+{
+	struct vxlan_fdb *f = NULL;
+	struct vxlan_rdst *rd = NULL;
+
+	rcu_read_lock();
+	f = __vxlan_find_mac(vxlan, mac, vni);
+	if (!f) {
+		rcu_read_unlock();
+		return -1;
+	}
+	list_for_each_entry(rd, &f->remotes, list) {
+		if (rd->remote_vni == vni) {
+			memcpy(rip, &rd->remote_ip, sizeof(union vxlan_addr));
+			rcu_read_unlock();
+			return 0;
+		}
+	}
+	rcu_read_unlock();
+	return -1;
+}
+EXPORT_SYMBOL_GPL(vxlan_find_remote_ip);
+
 /* Find and update age of fdb entry corresponding to MAC. */
 void vxlan_fdb_update_mac(struct vxlan_dev *vxlan, const u8 *mac, uint32_t vni)
 {
diff --git a/include/net/vxlan.h b/include/net/vxlan.h
index 52016af803c8..4b2a84163d46 100644
--- a/include/net/vxlan.h
+++ b/include/net/vxlan.h
@@ -363,6 +363,10 @@ struct vxlan_fdb_event {
 
 extern void vxlan_fdb_register_notify(struct notifier_block *nb);
 extern void vxlan_fdb_unregister_notify(struct notifier_block *nb);
+
+extern int vxlan_find_remote_ip(struct vxlan_dev *vxlan, const u8 *mac, __be32 vni,
+		                union vxlan_addr *rip);
+
 extern void vxlan_fdb_update_mac(struct vxlan_dev *vxlan, const u8 *mac, uint32_t vni);
 
 struct net_device *vxlan_dev_create(struct net *net, const char *name,
-- 
2.34.1

